applications/tari_validator_node/src/p2p/services/rpc_client.rs

```diff
@@ -0,0 +1,141 @@
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use async_trait::async_trait;
+use tari_common_types::types::PublicKey;
+use tari_comms::{
+    connection_manager::ConnectionManagerError,
+    connectivity::{ConnectivityError, ConnectivityRequester},
+    peer_manager::NodeId,
+    PeerConnection,
+};
+use tari_comms_dht::{envelope::NodeDestination, DhtDiscoveryRequester};
+use tari_crypto::tari_utilities::ByteArray;
+use tari_dan_core::{
+    models::TemplateId,
+    services::{ValidatorNodeClientFactory, ValidatorNodeRpcClient},
+    DigitalAssetError,
+};
+
+use crate::p2p::{proto::validator_node as proto, rpc};
+// const LOG_TARGET: &str = "tari::validator_node::p2p::services::rpc_client";
+
+pub struct TariCommsValidatorNodeRpcClient {
+    connectivity: ConnectivityRequester,
+    dht_discovery: DhtDiscoveryRequester,
+    address: PublicKey,
+}
+
+impl TariCommsValidatorNodeRpcClient {
+    async fn create_connection(&mut self) -> Result<PeerConnection, DigitalAssetError> {
+        if let Some(conn) = self
+            .connectivity
+            .get_connection(NodeId::from(self.address.clone()))
+            .await?
+        {
+            return Ok(conn);
+        }
+        match self.connectivity.dial_peer(NodeId::from(self.address.clone())).await {
+            Ok(connection) => Ok(connection),
+            Err(connectivity_error) => {
+                dbg!(&connectivity_error);
+                match &connectivity_error {
+                    ConnectivityError::ConnectionFailed(err) => {
+                        match err {
+                            ConnectionManagerError::PeerConnectionError(_) |
+                            ConnectionManagerError::DialConnectFailedAllAddresses |
+                            ConnectionManagerError::PeerIdentityNoValidAddresses => {
+                                // Try discover, then dial again
+                                self.dht_discovery
```

Comment:
TODO: Should discovery and connect the responsibility of the DHT layer

Reply:
Yeah agree